Index: qt-maemo-qtp/src/gui/kernel/qwidget.cpp
===================================================================
--- qt-maemo-qtp.orig/src/gui/kernel/qwidget.cpp
+++ qt-maemo-qtp/src/gui/kernel/qwidget.cpp
@@ -1326,7 +1326,7 @@
         data.crect = parentWidget ? QRect(0,0,100,30) : QRect(0,0,360,640);
     }
 #else
-    data.crect = parentWidget ? QRect(0,0,100,30) : QRect(0,0,640,480);
+    data.crect = parentWidget ? QRect(0,0,100,30) : QRect(0,0,854,480);
 #endif
 
     focus_next = focus_prev = q;
